src/darwin/Framework/CHIP/MTRAsyncCallbackWorkQueue.mm

```diff
@@ -169,9 +169,51 @@ - (void)_callNextReadyWorkItem
         self.runningWorkItemCount = 1;
 
         MTRAsyncCallbackQueueWorkItem * workItem = self.items.firstObject;
+
+        // Check if batching is possible or needed. Only ask work item to batch once for simplicity
+        if (workItem.batchable && workItem.batchingHandler && (workItem.retryCount == 0)) {
+            while (self.items.count >= 2) {
+                MTRAsyncCallbackQueueWorkItem * nextWorkItem = self.items[1];
+                if (!nextWorkItem.batchable || (nextWorkItem.batchingID != workItem.batchingID)) {
+                    // next item is not eligible to merge with this one
+                    break;
+                }
+
+                BOOL fullyMerged = NO;
+                workItem.batchingHandler(workItem.batchableData, nextWorkItem.batchableData, &fullyMerged);
+                MTR_LOG_DEFAULT(
+                    "MTRAsyncCallbackWorkQueue: merged items for batching - fully merged %@", fullyMerged ? @"YES" : @"NO");
```

Comment:
I just moved the logging to `MTRDevice.mm`, to include logPrefix. Small change.

Reply:
No, I mean ideally we would see the logPrefix bits from _both_ items in the logs about them being merged...

Reply:
I see... logPrefix would be the same, given the the match test is for the whole tuple of 3 ID's + params.

Reply:
For _batching_ it's not.

Reply:
Ah that's correct - I wasn't thinking right. In that case, I think the extra logging should be in the handler anyway, because the work queue/item system is not really supposed to know the details of the work items, for abstraction purpose. I could log more details when logging.

Reply:
Actually the merging details are already logged above:
```
                MTR_LOG_INFO("%@ batching merging %@ => %lu total", logPrefix, readRequestsNext[0],
                    (unsigned long) readRequestsCurrent.count);
```
where `readRequestsNext[0]` is the merged over read request (IDs + param).

And this "fully merged" log line is just for noting that a work item will be removed from the work queue.